src/components/organisms/UiMenu/UiMenu.vue

```diff
@@ -62,139 +53,175 @@ export interface MenuProps {
-      if (firstMenuItem.value && focusedMenuItem.value) {
-        focusedMenuItem.value.tabindex = -1;
-        firstMenuItem.value.tabindex = 0;
-        focusElement(firstMenuItem.value.$el.querySelector('.ui-button'));
-      }
+      handleMenuItemFocus(usedMenuTemplateRefs.value.at(0));
       break;
     case 'End':
     case 'PageDown':
-      if (lastMenuItem.value && focusedMenuItem.value) {
-        focusedMenuItem.value.tabindex = -1;
-        lastMenuItem.value.tabindex = 0;
-        focusElement(lastMenuItem.value.$el.querySelector('.ui-button'));
-      }
+      handleMenuItemFocus(usedMenuTemplateRefs.value.at(-1));
       break;
     case 'Tab':
-      const lastFocusedElement = ref(focusedMenuItem.value);
-      setTimeout(() => {
-        if(!lastFocusedElement.value) return;
-        lastFocusedElement.value.tabindex = -1;
-        if (selectedMenuItem.value) {
-          selectedMenuItem.value.tabindex = 0;
-          return;
-        }
-        if (firstMenuItem.value) {
-          firstMenuItem.value.tabindex = 0;
-        }
-      }, 0);
+      if (initialElement.value === activeElement) {
+        return;
+      }
+      if (activeElement) {
+        activeElement.tabindex = 0;
+        // We need to use setTimeout() because nextTick() does not bring the desired effect.
+        window.setTimeout(() => {
+          activeElement.tabindex = -1;
+        }, 0);
+      }
+      break;
+    default:
       break;
-    default: break;
   }
 };
-defineExpose({
-  menuItems,
-  firstMenuItem,
-  selectedMenuItem,
-});
-export interface MenuEmits {
-  (e: 'itemsLoaded'): void;
-}
-const emit = defineEmits<MenuEmits>();
-const hasMenuItems = computed(() => (
-  menuItems.value.length > 0
-));
-const setItemsNotReachable = () => {
-  [ ...menuItems.value ].forEach((item) => {
-    item.tabindex = item.$el.querySelector('.ui-button') ? -1 : 0;
-  });
-  if (selectedMenuItem.value) {
-    selectedMenuItem.value.tabindex = 0;
-  } else if(firstMenuItem.value) {
-    firstMenuItem.value.tabindex = 0;
+const isFocused = ref(false);
+const scrollLock = (event: KeyboardEvent) => {
+  if ((!props.enableKeyboardNavigation && !isFocused.value)
+      || !isFocused.value) {
+    return;
+  }
+  const { key } = event;
+  if ([
+    'ArrowUp',
+    'ArrowDown',
+    'Home',
+    'PageUp',
+    'End',
+    'PageDown',
+  ].includes(key)) {
+    event.preventDefault();
   }
-  emit('itemsLoaded');
 };
-watch(hasMenuItems, async (hasItems) => {
-  if (hasItems) {
-    await nextTick();
-    setItemsNotReachable();
+onMounted(() => {
+  window.addEventListener('keydown', scrollLock);
+});
+onBeforeUnmount(() => {
+  window.removeEventListener('keydown', scrollLock);
+});
+watch(focusedElement, (el, prevEl) => {
+  isFocused.value = !!(el && Object.keys(el).length > 0);
+  if (!el && prevEl) {
+    lastFocusedMenuItemTemplateRefs.value = prevEl;
+    if (initialElement.value) {
+      initialElement.value.tabindex = 0;
+    }
   }
 });
```

Comment:
And another comment about indentations 😅 Doesn't it look neater? 😄


```suggestion
watch(focusedElement, (el, prevEl) => {
  isFocused.value = !!(el && Object.keys(el).length > 0);
  if (!initialElement.value) return;
  if (el || !prevEl) return;
  initialElement.value.tabindex = 0;
  lastFocusedMenuItemTemplateRefs.value = prevEl;
});
```